Userland/Libraries/LibLocate/LocateDB.cpp

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright (c) 2021, Aron Lander <aron@aronlander.se>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <LibLocate/LocateDB.h>
+
+namespace Locate {
+
+LocateDB::LocateDB(String path, LocateDbMode mode)
+{
+    if (!Core::File::exists(locate_db_path)) {
+        auto file_handle = Core::File::open(locate_db_path, Core::IODevice::WriteOnly, 0640);
+        if (file_handle.is_error()) {
+            outln("Got error while trying to create locate db file: {}", file_handle.error());
+            VERIFY_NOT_REACHED();
+        }
+        file_handle.value()->close();
+        chown(locate_db_path, 0, 18);
+    }
+
+    m_file_handle = fopen(path.characters(), (mode == LocateDbMode::Write ? "wb" : "rb"));
+    fseek(m_file_handle, 0, SEEK_END);
+    m_file_size = ftell(m_file_handle);
+    fseek(m_file_handle, 0, SEEK_SET);
+}
+
+LocateDB::~LocateDB()
+{
+    fclose(m_file_handle);
+}
+
+void LocateDB::write_header()
+{
+    // Format: magic string + version.
+    String db_header = String::formatted("{}|{}", "locatedb", 1);
```

Comment:
Why does this need formatting, isn't this sufficient?

```suggestion
    String db_header("locatedb|1");
```